ksqldb-engine/src/main/java/io/confluent/ksql/services/SandboxedSchemaRegistryClient.java

```diff
@@ -36,21 +43,193 @@
 
   static SchemaRegistryClient createProxy(final SchemaRegistryClient delegate) {
     Objects.requireNonNull(delegate, "delegate");
-
-    return LimitedProxyBuilder.forClass(SchemaRegistryClient.class)
-        .swallow("register", anyParams(), 123)
-        .swallow("getId", anyParams(), 123)
-        .forward("getAllSubjects", methodParams(), delegate)
-        .forward("getSchemaById", methodParams(int.class), delegate)
-        .forward("getLatestSchemaMetadata", methodParams(String.class), delegate)
-        .forward("getSchemaBySubjectAndId", methodParams(String.class, int.class), delegate)
-        .forward("testCompatibility",
-            methodParams(String.class, ParsedSchema.class), delegate)
-        .swallow("deleteSubject", methodParams(String.class), Collections.emptyList())
-        .forward("getVersion", methodParams(String.class, ParsedSchema.class), delegate)
-        .build();
+    return new SandboxSchemaRegistryCache(delegate);
   }
 
   private SandboxedSchemaRegistryClient() {
   }
+
+  private static final class SandboxSchemaRegistryCache implements SchemaRegistryClient {
+    private final SchemaRegistryClient delegate;
+
+    private int nextSchemaId = Integer.MAX_VALUE;
+    private final Map<String, ParsedSchema> subjectCache = new HashMap<>();
+    private final Map<String, Integer> subjectToId = new HashMap<>();
+    private final Map<Integer, ParsedSchema> idCache = new HashMap<>();
+
+    private SandboxSchemaRegistryCache(final SchemaRegistryClient delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public Optional<ParsedSchema> parseSchema(
+        final String schemaType,
+        final String schemaString,
+        final List<SchemaReference> references) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int register(final String subject, final ParsedSchema parsedSchema) {
+      if (subjectCache.containsKey(subject)) {
+        throw new IllegalStateException("Subject '" + subject + "' already in use.");
+      }
+      final int schemaId = nextSchemaId--;
+      subjectCache.put(subject, parsedSchema);
+      subjectToId.put(subject, schemaId);
+      idCache.put(schemaId, parsedSchema);
```

Comment:
Could you use `MockSchemaRegistryClient`? https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/MockSchemaRegistryClient.java

Reply:
Maybe. What would we gain?

Reply:
1. Make maintaining the 3 caches here easier and less error prone? `MockSchemaRegistryClient` seems to have similar logic inside.
2. Make behavior same as prod schema registry. For example, `MockSchemaRegistryClient` could have register same subject differently instead of throwing.

Reply:
+1 if we can use the MockSchemaRegistryClient rather than maintaining our own caches for the newly registered subjects, that'd be nice. OTOH if it's a lot of effort I think the current approach is fine too. Let's just be careful about not throwing errors where we shouldn't be (see above).